core/src/main/java/org/apache/iceberg/TableScanContext.java

```diff
@@ -0,0 +1,146 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Collection;
+import java.util.Map;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * Context object with optional arguments for a TableScan.
+ */
+final class TableScanContext {
+  private final Long snapshotId;
+  private final Expression rowFilter;
+  private final boolean ignoreResiduals;
+  private final boolean caseSensitive;
+  private final boolean colStats;
+  private final Collection<String> selectedColumns;
+  private final ImmutableMap<String, String> options;
+  private final Long fromSnapshotId;
+  private final Long toSnapshotId;
+
+  TableScanContext() {
+    this(null, Expressions.alwaysTrue(), false, true, false, null, ImmutableMap.of(), null, null);
+  }
+
+  private TableScanContext(Long snapshotId, Expression rowFilter, boolean ignoreResiduals,
+                           boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+                           ImmutableMap<String, String> options, Long fromSnapshotId, Long toSnapshotId) {
+    this.snapshotId = snapshotId;
+    this.rowFilter = rowFilter;
+    this.ignoreResiduals = ignoreResiduals;
+    this.caseSensitive = caseSensitive;
+    this.colStats = colStats;
+    this.selectedColumns = selectedColumns;
+    this.options = options;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  Long snapshotId() {
+    return snapshotId;
+  }
+
+  TableScanContext snapshotId(Long id) {
+    return new TableScanContext(id, rowFilter, ignoreResiduals,
+        caseSensitive, colStats, selectedColumns, options, fromSnapshotId, toSnapshotId);
+  }
+
+  Expression rowFilter() {
+    return rowFilter;
+  }
+
+  TableScanContext rowFilter(Expression filter) {
+    return new TableScanContext(snapshotId, filter, ignoreResiduals,
+        caseSensitive, colStats, selectedColumns, options, fromSnapshotId, toSnapshotId);
+  }
+
+  boolean ignoreResiduals() {
+    return ignoreResiduals;
+  }
+
+  TableScanContext ignoreResiduals(boolean shouldIgnoreResiduals) {
+    return new TableScanContext(snapshotId, rowFilter, shouldIgnoreResiduals,
+        caseSensitive, colStats, selectedColumns, options, fromSnapshotId, toSnapshotId);
+  }
+
+  boolean caseSensitive() {
+    return caseSensitive;
+  }
+
+  TableScanContext caseSensitive(boolean isCaseSensitive) {
+    return new TableScanContext(snapshotId, rowFilter, ignoreResiduals,
+        isCaseSensitive, colStats, selectedColumns, options, fromSnapshotId, toSnapshotId);
+  }
+
+  boolean colStats() {
+    return colStats;
+  }
+
+  TableScanContext colStats(boolean shouldUseColumnStats) {
```

Comment:
This controls whether or not column stats are returned in the `DataFile` instances in the resulting splits, not whether column stats should be used in planning. Columns stats are always used in planning. To avoid confusion, can you rename this to `shouldReturnColumnStats`? A similar update for the getter would be useful as well.